### PR TITLE
Gavin/50 implement return search results api

### LINKED ISSUE #50: Implement Return Search Results API
## Feature Description
Implement the search API to return listings based on various criteria like keywords, category, price range, etc.

All of these Request Payload Parameters need to be implemented, supported, and tested:

- authorization (string, required): The user token for authentication.
- query (string, required): The search query text.
- minPrice (number, optional): Minimum price.
- maxPrice (number, optional): Maximum price.
- status (enum, optional): Filter by status - one of [SOLD, AVAILABLE] - default to AVAILABLE.
- searchType (enum, optional): Filter by users or listings - one of [USERS, LISTINGS] - default to LISTINGS. 
- latitude : (float, required): The latitude to search within 5 km of
- longitude: (float, required): The longitude to search within 5 km of
- sort (enum, optional): Sorting criteria - one of [RELEVANCE, PRICE_ASC, PRICE_DESC, LISTED_TIME_ASC, LISTED_TIME_DESC, DISTANCE_ASC, DISTANCE_DESC] (defaults to relevance).
- page (integer, required): Page number for pagination.
- limit (integer, required): Number of results per page.

It shall accept a valid sample request payload as stated in [SRS Section 5.1.4.1](https://docs.google.com/document/d/1DNqexMBf4MpYPCxWsT3plUDAjDSbdJwUzw-UyZX-vDs/edit#heading=h.dg3ntat152re), and return a corresponding success response with an array of (limit) listings.

It shall return a 400 Invalid Request error in the case that any parameters are missing or incorrectly formatted.

It shall additionally update the Postgres database with the search query for the user's search history purposes.

relates: nothing yet
blocked by: nothing yet

### User Stories To Implement

1. "As a user, I want to search for listings based on certain keywords and criteria, so that I can find items to buy."
2. "As a user, I want to search for sellers, so that I can find specific users and view their listings."
3. "As a registered user, I want to be able to view my search history, so that I can revisit previous searches or items."

## Related External Resources
[SRS Section 5.1.4.1](https://docs.google.com/document/d/1DNqexMBf4MpYPCxWsT3plUDAjDSbdJwUzw-UyZX-vDs/edit#heading=h.dg3ntat152re)
#189

## Acceptance Criteria

Sorted by User story

### User Story 1: "As a user, I want to search for listings based on certain keywords and criteria, so that I can find items to buy."

**Unit Tests:**
1. **Filter Application:**
   - Test that the search query filters by keywords correctly.
   - Test that the price range filter (minPrice and maxPrice) is applied correctly.
   - Test that the status filter (AVAILABLE, SOLD) is applied correctly.
   - Test that the location filter (latitude, longitude) is applied correctly within a 5 km radius.
   
2. **Sorting Functionality:**
   - Test that the sorting by relevance is applied correctly.
   - Test that the sorting by price (ascending and descending) is applied correctly.
   - Test that the sorting by listed time (ascending and descending) is applied correctly.
   - Test that the sorting by distance (ascending and descending) is applied correctly.

3. **Pagination:**
   - Test that the pagination works correctly by retrieving the correct number of results per page.
   - Test that the correct page number is returned for multiple pages.

**Acceptance Tests:**
1. **Search Results:**
   - Given a valid search query, when the user searches for listings, then the system should return listings that match the search criteria.
   - Given a search query with a specified price range, when the user searches, then the system should return listings within the specified price range.
   - Given a search query with a status filter, when the user searches, then the system should return listings with the specified status.
   - Given a search query with a location filter, when the user searches, then the system should return listings within 5 km of the specified location.

2. **Sorting:**
   - Given a search query, when the user sorts by relevance, then the system should return listings sorted by relevance.
   - Given a search query, when the user sorts by price (ascending/descending), then the system should return listings sorted by price.
   - Given a search query, when the user sorts by listed time (ascending/descending), then the system should return listings sorted by listed time.
   - Given a search query, when the user sorts by distance (ascending/descending), then the system should return listings sorted by distance.

3. **Pagination:**
   - Given a search query, when the user requests a specific page, then the system should return the correct listings for that page.
   - Given a search query, when the user specifies a limit of results per page, then the system should return the specified number of listings per page.

### User Story 2: "As a user, I want to search for sellers, so that I can find specific users and view their listings."

**Unit Tests:**
1. **User Search:**
   - Test that the search query filters by seller name correctly when searchType is input as USERS.
   
2. **Listings Retrieval:**
   - Test that the listings associated with a seller are retrieved correctly.

**Acceptance Tests:**

1. **Seller Listings:**
   - Given a valid search query, when the user searches for a specific seller, then the system should display the listings associated with that seller.

### User Story 3: "As a registered user, I want to be able to view my search history, so that I can revisit previous searches or items."

**Unit Tests:**
1. **Search History Storage:**
   - Test that a search query is stored in the PostgreSQL database correctly (most recent first).
   - Test that the search history is associated with the correct user.

**Acceptance Tests:**
1. **Search History Storage:**
   - Given a registered user, when the user performs a search, then the system should store the search query in the PostgreSQL database associated with the user.
   - Given a registered user, when the user performs multiple searches, then the system should store each search query in the PostgreSQL database associated with the user.

2. **Search History Retrieval:**
   - Given a registered user, when the user views their search history, then the system should retrieve and display the search history from the PostgreSQL database.
   - Given a registered user, when the user views their search history, then the system should display the search queries in the correct order (most recent first).

### PR DESCRIPTION
**Pull request by GDeane**
# Description
This is a PR showing the return search results API and its corresponding tests.

### It Includes:
1. All of these Request Payload Parameters implemented, supported, and tested:
- authorization (string, required): The user token for authentication.
- query (string, required): The search query text.
- minPrice (number, optional): Minimum price.
- maxPrice (number, optional): Maximum price.
- status (enum, optional): Filter by status - one of [SOLD, AVAILABLE] - default to AVAILABLE.
- searchType (enum, optional): Filter by users or listings - one of [USERS, LISTINGS] - default to LISTINGS.
- latitude : (float, required): The latitude to search within 5 km of
- longitude: (float, required): The longitude to search within 5 km of
- sort (enum, optional): Sorting criteria - one of [RELEVANCE, PRICE_ASC, PRICE_DESC, LISTED_TIME_ASC, LISTED_TIME_DESC, DISTANCE_ASC, DISTANCE_DESC] (defaults to relevance).
- page (integer, optional): Page number for pagination.
- limit (integer, optional): Number of results per page.

2. Pydantic BaseModels

3. Integration tests using a test-index which is initialized before each test and destroyed after each test. All tests use the elasticsearch endpoint.

### It **DOES NOT INCLUDE**
1. Dealing with the authorization cookie (@MNThomson) was going to set that up on Tuesday if it wasn't done by then. For now there is no use for the userID (that will be needed for dealing with search history)

2. Updating the user_searches table for search history purposes

Closes #50 

### To run automatic tests:
1. Checkout this branch
3. In a terminal, navigate to the martletplace directory and run `docker-compose up --build`
4. Wait a minute for the data layer (particularly elasticsearch) to get ready for input
5. In a terminal, run `docker exec -it martletplace_search bash`
6. In the bash terminal, run `pytest src/test_server.py`

## Checklist
- [X] The code includes tests if relevant
- [x] I have *actually* self-reviewed my changes and done QA

## Additional Notes
1. As seen in test_server.py, it is important that the location (latitude and longitude) is set up as a geo_point for elasticsearch to work. It is also important that the parameters passed are named "lat" and "lon" as opposed to "latitude" and "longitude" (this is relevant to @emilasuska for implementing the reindexing methods)
image
image

2. I am unsure about some of the error codes that are used. The reason I have used code 422 (Unprocessable Content) a lot is because FastAPI uses it a lot for its built-in error handling.  Please call me out for anything that should be changed though

### CODE REVIEW

File: apps/algorithm/search/src/server.py [hunk 1]
```diff
@@ -65,35 +129,99 @@ async def search(
...
+                {"multi_match": {"query": query, "fields": ["title", "description"]}},
+                {"match": {"status": status}},
+            ]
+        elif searchType == "USERS":
+            must_conditions = [
+                {"match": {"sellerName": query}},
+                {"match": {"status": status}},
+            ]
+        else:
+            raise HTTPException(status_code=400, detail="Invalid searchType")
```
Comment (Scott-Kenning) [inline apps/algorithm/search/src/server.py:162]: error codes should be consistent, generally 400 is used for a missing or incorrect parameter, and 422 is used for a syntactically correct parameter that fails business rules. In this case there's a valid searchType field with a bad value, so it should return 422

**Review by Scott-Kenning — APPROVED** [review]
LGTM, don't see any major issues. I would recommend factoring out your code more into different files so each file isn't so big, I didn't review test_server because its just test code and its 1300 lines

File: apps/algorithm/search/src/server.py (hunk 1 above)
Comment (GDeane) [reply apps/algorithm/search/src/server.py]: did it

File: apps/algorithm/search/src/server.py [hunk 2]
```diff
@@ -65,35 +129,99 @@ async def search(
...
+
+        sort_options = {
+            "RELEVANCE": "_score",
+            "PRICE_ASC": "price",
+            "PRICE_DESC": "price",
+            "LISTED_TIME_ASC": "dateCreated",
+            "LISTED_TIME_DESC": "dateCreated",
+            "DISTANCE_ASC": "_geo_distance",
+            "DISTANCE_DESC": "_geo_distance",
+        }
```
Comment (MNThomson) [inline apps/algorithm/search/src/server.py:149]: If this is only used in one place why is it top level and super far away from its use? Let's move down to right above where it's used (locality of behaviour)

File: apps/algorithm/search/src/server.py [hunk 3]
```diff
@@ -65,35 +129,99 @@ async def search(
...
+            if maxPrice is not None:
+                price_range["lte"] = maxPrice
+            search_body["query"]["bool"]["filter"].append(
+                {"range": {"price": price_range}}
+            )
+
+        search_body["query"]["bool"]["filter"].append(
+            {
+                "geo_distance": {
+                    "distance": "5km",
```
Comment (MNThomson) [inline apps/algorithm/search/src/server.py:184]: Should this be hard-coded or perhaps a constant?

File: apps/algorithm/search/src/test_server.py [hunk 4]
```diff
@@ -1,2 +1,1323 @@
...
+
+es_endpoint = os.getenv("ES_ENDPOINT")
+es = Elasticsearch([es_endpoint], verify_certs=False)
+
+
+@pytest.fixture(scope="function", autouse=True)
+def setup_and_teardown_index(monkeypatch):
+    monkeypatch.setenv("ES_INDEX", TEST_INDEX)
+
+    es.options(ignore_status=[400]).indices.create(
```
Comment (MNThomson) [inline apps/algorithm/search/src/test_server.py:21]: Why are we ignoring `400` for all the tests?

File: apps/algorithm/search/src/test_server.py [hunk 5]
```diff
@@ -1,2 +1,1323 @@
...
+
+client = TestClient(app)
+
+es_endpoint = os.getenv("ES_ENDPOINT")
+es = Elasticsearch([es_endpoint], verify_certs=False)
+
+
+@pytest.fixture(scope="function", autouse=True)
+def setup_and_teardown_index(monkeypatch):
+    monkeypatch.setenv("ES_INDEX", TEST_INDEX)
```
Comment (MNThomson) [inline apps/algorithm/search/src/test_server.py:19]: We talked about using a different "random" Index for each test for isolation. If we're switching to async tests (for postgres connections), this'll be required

File: apps/algorithm/search/src/server.py [hunk 6]
```diff
@@ -65,35 +129,99 @@ async def search(
...
+                        "unit": "km",
+                    }
+                }
+            )
+        else:
+            search_body["sort"].append(
+                {sort_options[sort]: {"order": "asc" if "ASC" in sort else "desc"}}
+            )
+
+        response = es.search(index=INDEX, body=search_body)
```
[the remark this answers is not in the document]
Comment (GDeane) [reply apps/algorithm/search/src/server.py:205]: Will do

File: apps/algorithm/search/src/server.py (hunk 3 above)
Comment (GDeane) [reply apps/algorithm/search/src/server.py:184]: Made it a constant for easy changing if we want to later

File: apps/algorithm/search/src/test_server.py (hunk 5 above)
Comment (GDeane) [reply apps/algorithm/search/src/test_server.py:19]: I'll do that in a PR when I add the postgres stuff and switch to async tests.

File: apps/algorithm/search/src/test_server.py (hunk 4 above)
Comment (GDeane) [reply apps/algorithm/search/src/test_server.py:21]: Initially I didn't want the test to fail if an index already existed, but on second thought that could lead to unpredictable state problems, so I don't want to do that.

removing that.

**Comment by GDeane** [comment]
I still need to update based on the API spec requirement to also return the total number of search hits for pagination display reasons.

**Review by emilasuska — APPROVED** [review]

**Pull request merged**